pkg/apb/ext_creds.go

```diff
@@ -10,82 +10,68 @@ import (
-// output is *very* experimental and error prone. Entire approach is going
-// to be thrown out and redone asap.
 func ExtractCredentials(
 	podname string, namespace string, log *logging.Logger,
 ) (*ExtractedCredentials, error) {
-	log.Debug("Calling monitorOutput on " + podname)
-	credOut, _ := monitorOutput(podname, namespace)
-	log.Debug("oc log output: %s", string(credOut))
+	credOut := make(chan []byte)
 
-	var creds *ExtractedCredentials
-	var err error
-	creds, err = buildExtractedCredentials(credOut)
+	log.Debug("Calling monitorOutput on " + podname)
+	go monitorOutput(namespace, podname, credOut, log)
 
-	if err != nil {
-		// HACK: this is HORRIBLE. but there is definitely a time between a bind
-		// and when the container is up.
-		totalRetries := totalTimeout / timeoutFreq
-		retries := 1
-		for {
-			if retries == totalRetries {
-				errstr := "TIMED OUT WAITING FOR CONTAINER TO COME UP"
-				log.Error(errstr)
-				return nil, errors.New(errstr)
-			}
-
-			time.Sleep(time.Duration(timeoutFreq) * time.Second)
-			log.Info("Container not up yet, retrying %d of %d on pod %s", retries, totalRetries, podname)
-			credOut, _ = monitorOutput(podname, namespace)
-			log.Debug("oc log output: \n%s", string(credOut))
-			creds, err = buildExtractedCredentials(credOut)
-
-			if err == nil {
-				if creds != nil {
-					log.Debug("Pod reporting finished and returned Credentials")
-				} else {
-					log.Debug("Pod reporting finished and DID NOT return Credentials")
-				}
-				break
-			} else if err.Error() == stillWaitingError {
-				// Known error code that's received when we're either waiting for
-				// ContainerCreating, or for the pod resource to be created.
-				// These are expected states, and we'll wait until the pod is up.
-				log.Debug(err.Error())
-			} else {
-				log.Notice("WARNING: Unexpected output from apb pod")
-				log.Notice("Will keep retrying, but it's possible something has gone wrong.")
-				log.Notice(err.Error())
-			}
-
-			retries++
-		}
+	var bindOutput []byte
+	for msg := range credOut {
+		bindOutput = msg
 	}
 
-	return creds, err
-}
+	if bindOutput == nil {
+		return nil, nil
+	}
 
-func monitorOutput(podname string, namespace string) ([]byte, error) {
-	return RunCommand("oc", "logs", "-f", "--namespace="+namespace, podname)
+	return buildExtractedCredentials(bindOutput)
 }
 
-func buildExtractedCredentials(output []byte) (*ExtractedCredentials, error) {
-	stillWaiting := strings.Contains(string(output), "ContainerCreating") ||
-		strings.Contains(string(output), "NotFound")
+func monitorOutput(namespace string, podname string, mon chan []byte, log *logging.Logger) {
+	// TODO: Error handling here
+	// It would also be nice to gather the script output that exec runs
+	// instead of only getting the credentials
+
+	for r := 1; r <= CredentialRetries; r++ {
+		output, _ := RunCommand("oc", "exec", podname, GatherCredentialsCMD, "--namespace="+namespace)
```

Comment:
💯 this will totally bite us down the road. I can already foresee the exec erroring out for some odd reason and not having any idea what really happened.